### PR TITLE
Feature/arc 1425 object to csv

### PR DESCRIPTION
**Pull request by EmileVanTichelenSHD**
https://meemoo.atlassian.net/browse/ARC-1425

New packages: papaparse, flat

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -52,6 +52,7 @@
     "date-fns": "2.28.0",
     "date-fns-tz": "1.3.0",
     "express-session": "1.17.2",
+    "flat": "^5.0.2",
```
Comment (tourlbr) [inline package.json:55]: REQ: remove flat npm package

File: src/modules/ie-objects/controllers/ie-objects.controller.ts [hunk 2]
```diff
@@ -144,6 +145,44 @@ export class IeObjectsController {
...
+				time: new Date().toISOString(),
+			},
+		]);
+
+		// Limit access to the objects in the collection
+		const visitorSpaceAccessInfo =
+			await this.ieObjectsService.getVisitorSpaceAccessInfoFromUser(user);
+
+		return convertObjectToCsv(
+			limitAccessToObjectDetails(objectMetadata, {
```
Comment (tourlbr) [inline src/modules/ie-objects/controllers/ie-objects.controller.ts:174]: REQ: both csv and xml should not be using **limitAccessToObjectDetails**

File: src/modules/ie-objects/helpers/convert-objects-to-csv.ts [hunk 3]
```diff
@@ -0,0 +1,14 @@
+import { flatten } from 'flat';
+import Papa from 'papaparse';
+
+import { IeObject } from '../ie-objects.types';
+
+export const convertObjectsToCsv = (objects: Partial<IeObject>[]): string => {
```
Comment (tourlbr) [inline src/modules/ie-objects/helpers/convert-objects-to-csv.ts:6]: REQ: test file needs to be written with several usecase to be sure the code works as intended. Example can be found in convert to xml spec file.

File: src/modules/ie-objects/helpers/convert-objects-to-xml.ts [hunk 4]
```diff
@@ -5,9 +5,11 @@ import { IeObject } from '../ie-objects.types';
 export const convertObjectsToXml = (objects: Partial<IeObject>[]): string => {
```
Comment (tourlbr) [inline src/modules/ie-objects/helpers/convert-objects-to-xml.ts:5]: Q: was the spec file reviewed after these code changes to be sure they still succeed?
REQ: review test case and update were necessary

**Review by tourlbr — CHANGES_REQUESTED** [review]
Don't forget to write tests for the code you have written.

File: src/modules/ie-objects/helpers/convert-objects-to-csv.ts [hunk 5]
```diff
@@ -0,0 +1,17 @@
+import { flatten } from 'flat'; //toch flat gebruiken
```
Comment (tourlbr) [inline src/modules/ie-objects/helpers/convert-objects-to-csv.ts:1]: REQ: remove comment

File: src/modules/ie-objects/helpers/convert-objects-to-xml.ts [hunk 6]
```diff
@@ -1,13 +1,20 @@
...
 	// this structure defines the parent 'objects' tag, which includes
 	// all objects wrapped in a separate 'object' tag
-	return convert.js2xml({ objects: { object: objects } }, { compact: true, spaces: 2 });
+	const xmlData = objects.map((o) => _.pick(o, IE_OBJECT_PROPS_METADATA_EXPORT));
+	return convert.js2xml({ objects: { object: xmlData } }, { compact: true, spaces: 2 });
 };
 
 export const convertObjectToXml = (object: Partial<IeObject>): string => {
-	return convert.js2xml({ object }, { compact: true, spaces: 2 });
+	// const { accessThrough, ...xmlData } = object;
```
Comment (tourlbr) [inline src/modules/ie-objects/helpers/convert-objects-to-xml.ts:15]: REQ: remove unused line

File: src/modules/ie-objects/services/ie-objects.service.ts [hunk 7]
```diff
@@ -231,7 +231,7 @@ export class IeObjectsService {
 	}
 
 	/**
-	 * Get the object detail fields that are exposed as metadata
+	 * Get the objeµt detail fields that are exposed as metadata
```
Comment (tourlbr) [inline src/modules/ie-objects/services/ie-objects.service.ts:234]: REQ: fix typo

**Review by tourlbr — APPROVED** [review]
I'm going to pre approve this but please the necessary code based on the comments I left behind.

File: src/modules/ie-objects/ie-objects.conts.ts [hunk 8]
```diff
@@ -243,3 +243,44 @@ export const IE_OBJECT_PROPS_BY_METADATA_SET: Record<string, string[]> = {
...
+	'inLanguage',
+	//meemoo_description_categorie => Not currently available in DB
+	'dctermsMedium',
+	'dctermsFormat',
+	'meemoofilmColor',
+	'meemoofilmBase',
+	'meemoofilmImageOrSound',
+	//meemoo_contains_embedded_caption => Not currently available in DB
+	//language_subtitles => Not currently available in DB
+	//ebucore_is_media_fragment_of => Not currently available in DB
```
Comment (bertyhell) [inline src/modules/ie-objects/ie-objects.conts.ts:285]: These values are indeed not available yet, but i don't think the code breaks if you enable these, right?
If so we should enable them, that way we don't have to touch this code anymore once those fields become available and the export will just work as expected.

File: src/modules/ie-objects/helpers/convert-objects-to-csv.spec.ts [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+
+	it('returns the csv version of an array of nested objects', () => {
+		const csv = convertObjectsToCsv([
+			{
+				otherKey: 'otherValue',
+				meemooOriginalCp: '1',
+				creator: { Maker: 'test' },
+			},
+		] as unknown as IeObject[]);
+		console.log(csv);
```
Comment (bertyhell) [inline src/modules/ie-objects/helpers/convert-objects-to-csv.spec.ts:41]: Remove this console.log

File: package.json [hunk 10]
```diff
@@ -61,7 +62,9 @@
     "lodash": "4.17.21",
     "node-cron": "3.0.0",
     "node-zendesk": "^2.2.0",
+    "papaparse": "^5.4.0",
     "query-string": "7.1.0",
+    "ramda": "^0.28.0",
```
Comment (bertyhell) [inline package.json:67]: don't add ramda since we already use lodash in this repository

**Review by bertyhell — CHANGES_REQUESTED** [review]

**Review by bertyhell — APPROVED** [review]

**Pull request merged**